File: java/creational/MultitonPattern.java
package creational;

import java.util.HashMap;
import java.util.Map;

/**
 * Multiton Pattern
 * Ensures only one instance per key exists.
 */
public class MultitonPattern {
    
    // Multiton class
    static class DatabaseConnection {
        private static final Map<String, DatabaseConnection> instances = new HashMap<>();
        private final String database;
        
        private DatabaseConnection(String database) {
            this.database = database;
            System.out.println("Creating connection to database: " + database);
        }
        
        public static synchronized DatabaseConnection getInstance(String key) {
            if (!instances.containsKey(key)) {
                instances.put(key, new DatabaseConnection(key));
            } else {
                System.out.println("Reusing existing connection to: " + key);
            }
            return instances.get(key);
        }
        
        public void executeQuery(String query) {
            System.out.println("Executing on " + database + ": " + query);
        }
        
        public static void printAllInstances() {
            System.out.println("\nTotal instances: " + instances.size());
            instances.keySet().forEach(key -> 
                System.out.println("  - " + key)
            );
        }
    }
    
    // Another Multiton example - Logger
    static class Logger {
        private static final Map<String, Logger> instances = new HashMap<>();
        private final String name;
        
        private Logger(String name) {
            this.name = name;
        }
        
        public static synchronized Logger getInstance(String name) {
            return instances.computeIfAbsent(name, Logger::new);
        }
        
        public void log(String message) {
            System.out.println("[" + name + "] " + message);
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=== Multiton Pattern Demo ===\n");
        
        System.out.println("1. Database Connection Multiton:");
        
        // Get instances for different databases
        DatabaseConnection mysql1 = DatabaseConnection.getInstance("MySQL");
        mysql1.executeQuery("SELECT * FROM users");
        
        DatabaseConnection postgres1 = DatabaseConnection.getInstance("PostgreSQL");
        postgres1.executeQuery("SELECT * FROM products");
        
        DatabaseConnection mongo1 = DatabaseConnection.getInstance("MongoDB");
        mongo1.executeQuery("db.orders.find()");
        
        // Try to get instances again - should reuse existing
        System.out.println("\nRequesting existing instances:");
        DatabaseConnection mysql2 = DatabaseConnection.getInstance("MySQL");
        DatabaseConnection postgres2 = DatabaseConnection.getInstance("PostgreSQL");
        
        // Verify same instances
        System.out.println("\nVerifying instances:");
        System.out.println("mysql1 == mysql2: " + (mysql1 == mysql2));
        System.out.println("postgres1 == postgres2: " + (postgres1 == postgres2));
        
        DatabaseConnection.printAllInstances();
        
        // Logger example
        System.out.println("\n\n2. Logger Multiton:");
        Logger appLogger = Logger.getInstance("Application");
        Logger dbLogger = Logger.getInstance("Database");
        Logger secLogger = Logger.getInstance("Security");
        
        appLogger.log("Application started");
        dbLogger.log("Connected to database");
        secLogger.log("User authenticated");
        appLogger.log("Processing request");
        dbLogger.log("Query executed");
        
        // Reuse logger
        Logger appLogger2 = Logger.getInstance("Application");
        System.out.println("\nappLogger == appLogger2: " + (appLogger == appLogger2));
        
        System.out.println("\n--- Key Differences from Singleton ---");
        System.out.println("Singleton: One instance per class");
        System.out.println("Multiton: One instance per key per class");
    }
}

File: python/creational/multiton_pattern.py
"""
Multiton Pattern
Purpose: Ensures only one instance per key exists
Use Case: Managing multiple named instances, registry pattern
"""

from typing import Dict, Optional
import threading


class Multiton:
    _instances: Dict[str, 'Multiton'] = {}
    _lock = threading.Lock()
    
    def __new__(cls, key: str):
        if key not in cls._instances:
            with cls._lock:
                if key not in cls._instances:
                    instance = super().__new__(cls)
                    cls._instances[key] = instance
        return cls._instances[key]
    
    def __init__(self, key: str):
        if not hasattr(self, '_initialized'):
            self.key = key
            self._initialized = True


class DatabaseConnection(Multiton):
    """Database connection pool with multiton"""
    def __init__(self, key: str):
        super().__init__(key)
        if not hasattr(self, 'connection_string'):
            self.connection_string = f"Connection to {key}"
    
    def query(self, sql: str) -> str:
        return f"[{self.key}] Executing: {sql}"


class CacheManager:
    """Cache manager with multiton pattern"""
    _instances: Dict[str, 'CacheManager'] = {}
    _lock = threading.Lock()
    
    @classmethod
    def get_instance(cls, region: str) -> 'CacheManager':
        if region not in cls._instances:
            with cls._lock:
                if region not in cls._instances:
                    cls._instances[region] = cls(region)
        return cls._instances[region]
    
    def __init__(self, region: str):
        self.region = region
        self.cache: Dict[str, any] = {}
    
    def set(self, key: str, value: any):
        self.cache[key] = value
    
    def get(self, key: str) -> Optional[any]:
        return self.cache.get(key)


if __name__ == "__main__":
    print("=== Multiton Pattern Demo ===\n")
    
    # Database connections
    print("--- Database Connections ---")
    db_primary = DatabaseConnection("primary")
    db_replica = DatabaseConnection("replica")
    db_primary2 = DatabaseConnection("primary")
    
    print(f"db_primary is db_primary2: {db_primary is db_primary2}")
    print(f"db_primary is db_replica: {db_primary is db_replica}")
    print(db_primary.query("SELECT * FROM users"))
    print(db_replica.query("SELECT * FROM users"))
    
    # Cache managers
    print("\n--- Cache Managers ---")
    cache_us = CacheManager.get_instance("us-east-1")
    cache_eu = CacheManager.get_instance("eu-west-1")
    cache_us2 = CacheManager.get_instance("us-east-1")
    
    print(f"cache_us is cache_us2: {cache_us is cache_us2}")
    print(f"cache_us is cache_eu: {cache_us is cache_eu}")
    
    cache_us.set("user:1", {"name": "Alice"})
    print(f"US Cache: {cache_us2.get('user:1')}")
    print(f"EU Cache: {cache_eu.get('user:1')}")
